### PR TITLE
01) defined the struct for each of my 10 types

### PR DESCRIPTION
**Pull request by MartinMueller-StudioMood**
02) declared member variables and member functions
03) gave the member variables relevant data types
04) made the function parameter list for those member functions and used some of your User-Defined Types
05) 2 of the 10 user-defined types have a nested class
06) 10th UDTs should only use 5 of my other UDTs for its member variable types

### CODE REVIEW

File: main.cpp [hunk 1]
```diff
@@ -171,6 +288,33 @@ struct CarWash //                                   1) define an empty struct fo
...
+    int bassdrum = 12; //3) member variables with relevant data types.
+    //has snare
+    int snare = 8; //3) member variables with relevant data types.
+    //has tom
+    int tom = 4; //3) member variables with relevant data types.
+    //has hihat
+    int hihat = 5; //3) member variables with relevant data types.
+    //has sequencer
+    int sequencer = 127; //3) member variables with relevant data types.
+
```
Comment (matkatmusic) [inline main.cpp:303]: What are these, midi note numbers?   What kind of information are these properties supposed to represent?
- [ ] revise.  pick properties that provide information about your `DrumMachine`

File: main.cpp [hunk 2]
```diff
@@ -185,10 +329,38 @@ struct CarWash //                                   1) define an empty struct fo
...
+    //has volume
+    int volume = 127; //3) member variables with relevant data types.
+    //has sin
+    int sin = 127; //3) member variables with relevant data types.
+    //has square
+    int square = 23; //3) member variables with relevant data types.
+    //has triangle
+    int triangle = 27; //3) member variables with relevant data types.
+    //has noise
+    int noise = 127; //3) member variables with relevant data types.
```
Comment (matkatmusic) [inline main.cpp]: What kind of information are these properties supposed to represent?
- [ ] revise.  pick properties that provide information and name them accordingly.

File: main.cpp [hunk 3]
```diff
@@ -199,12 +371,39 @@ struct CarWash //                                   1) define an empty struct fo
...
+    //has attack
+    int attack = 127; //3) member variables with relevant data types.
+    //has decay
+    int decay = 127; //3) member variables with relevant data types.
+    //has sustain
+    int sustain = 23; //3) member variables with relevant data types.
+    //has release
+    int release = 27; //3) member variables with relevant data types.
+    //has amount
+    int amount = 127; //3) member variables with relevant data types.
```
Comment (matkatmusic) [inline main.cpp]: What kind of information are these properties supposed to represent?
- [ ] revise.  pick properties that provide information and name them accordingly.

File: main.cpp [hunk 4]
```diff
@@ -199,12 +371,39 @@ struct CarWash //                                   1) define an empty struct fo
...
+    member functions with a user-defined type as the parameter.
+    The user-defined type parameter happens to be the nested class.
+    */
+
+    //2) define attack of a sound
+    void changeAttack();
+    //2) decay a sound
+    void changeDecay();
+    //2) define the release of a sound
+    void changeRelease();
```
Comment (matkatmusic) [inline main.cpp]: These definitely need function parameters since something is being changed.

File: main.cpp [hunk 5]
```diff
@@ -213,6 +412,34 @@ struct CarWash //                                   1) define an empty struct fo
...
+    //has value
+    int value = 127; //3) member variables with relevant data types.
+    //has channel
+    int channel = 127; //3) member variables with relevant data types.
+    //has input
+    int input = 127; //3) member variables with relevant data types.
+    //has output
+    int output = 127; //3) member variables with relevant data types.
+    //has sync type
+    int syncType = 127; //3) member variables with relevant data types.
```
Comment (matkatmusic) [inline main.cpp]: What kind of information are these properties supposed to represent?
- [ ] revise.  pick properties that provide information and name them accordingly.

File: main.cpp [hunk 6]
```diff
@@ -213,6 +412,34 @@ struct CarWash //                                   1) define an empty struct fo
...
+    member functions with a user-defined type as the parameter.
+    The user-defined type parameter happens to be the nested class.
+    */
+
+    //2) send midi
+    void sendMidi();
+    //2) receive midi
+    void receiveMidi();
+    //2) controll instruments
+    void controllInstrument();
```
Comment (matkatmusic) [inline main.cpp]: `Midi::sendMidi()`...   What you want is `MidiInterface::sendMidi(MidiMessage)`
that means 2 classes (here's a chance for your nested class) are needed here.
- [ ] revise

File: main.cpp [hunk 7]
```diff
@@ -227,6 +454,32 @@ struct CarWash //                                   1) define an empty struct fo
...
+    //has rate
+    int rate = 127; //3) member variables with relevant data types.
+    //has amount
+    int amount = 127; //3) member variables with relevant data types.
+    //has waveform
+    int waveform = 127; //3) member variables with relevant data types.
+    //has adsr
+    int adsr = 127; //3) member variables with relevant data types.
+    //has retrigger
+    int retrigger = 127; //3) member variables with relevant data types.
```
Comment (matkatmusic) [inline main.cpp]: What kind of information are these properties supposed to represent?
this is obviously just a copy/paste/rename.  Be more creative.
- [ ] revise.  pick properties that provide information and name them accordingly.

File: main.cpp [hunk 8]
```diff
@@ -227,6 +454,32 @@ struct CarWash //                                   1) define an empty struct fo
...
+    member functions with a user-defined type as the parameter.
+    The user-defined type parameter happens to be the nested class.
+    */
+
+    //2) create waveform
+    void createWaveform();
+    //2) controll other parameters with waveform
+    void sendWaveform();
+    //2) define waveform with adsr
+    void defineWaveformAdsr();
```
Comment (matkatmusic) [inline main.cpp]: Definitely need parameters here.   what waveform is being created?   Where is the waveform being sent?   
What are the ADSR parameters that are being used to create the waveform ADSR? 
- [ ] revise

File: main.cpp [hunk 9]
```diff
@@ -241,6 +494,32 @@ struct CarWash //                                   1) define an empty struct fo
...
+    //has sample slot
+    int sampleSlot = 127; //3) member variables with relevant data types.
+    //has start
+    int start = 127; //3) member variables with relevant data types.
+    //has loop
+    int loop = 127; //3) member variables with relevant data types.
+    //has lenght
+    int lenght = 127; //3) member variables with relevant data types.
+    //has fade
+    int fade = 127; //3) member variables with relevant data types.
```
Comment (matkatmusic) [inline main.cpp]: What kind of information are these properties supposed to represent?
again with the copy/paste/rename.   Be more explicit when you name your properties and choose your property types so they convey some actual information.   `startSample`  `loopLength`  `isLooped` `fadeLengthInSamples`  or `fadeLengthInMilliseconds` for example.  Help the reader understand what your code is supposed to do.
- [ ] revise.  pick properties that provide information and name them accordingly.

File: main.cpp [hunk 10]
```diff
@@ -270,8 +550,36 @@ Thing 10) Synthesizer
...
+    member functions with a user-defined type as the parameter.
+    The user-defined type parameter happens to be the nested class.
+    */
+
+    //2) create sounds
+    void createSounds();
+    //2) send midi
+    void sendMidi();
+    //2) manipulate the sound with adsr evelopes
+    void manipulateSoundAdsr();
```
Comment (matkatmusic) [inline main.cpp:589]: - [ ] add params that are related to what these functions are supposed to do based on their description.

**Review by matkatmusic — CHANGES_REQUESTED** [review]
Most every UDT uses `int` for the property type, and `void` for the function return type.
change it up.  make your properties use other types.  make your functions return something other than void. 
- [ ] revise.

File: main.cpp (hunk 1 above)
Comment (martin-mueller-xyz) [reply main.cpp:303]: this is the index of the different bassdrums my instrument provides. i have 12 different bassdrum samples for example and the user can choose which bassdrum he want to play. it's the same with the other instrumentents.

File: main.cpp (hunk 2 above)
Comment (martin-mueller-xyz) [reply main.cpp]: These are the properties of my oscillator.

File: main.cpp (hunk 3 above)
Comment (martin-mueller-xyz) [reply main.cpp]: this is my adsr to provide the following properties to the adsr section. int attack is the value of the attack parameter for example.

File: main.cpp (hunk 5 above)
Comment (martin-mueller-xyz) [reply main.cpp]: this is my midi function to play my instrument via midi notes.

File: main.cpp (hunk 7 above)
Comment (martin-mueller-xyz) [reply main.cpp]: this is the lfo with his values of my instrument

**Review by matkatmusic — APPROVED** [review]

**Pull request merged**